### ISSUE TITLE
Parsing embedded JSON

### ISSUE THREAD

**Issue by geekscrapy** [rcoh/angle-grinder#64]
Hi,

I have the following JSON format:

{"dst_host": "159.65.224.130", "dst_port": 23, "honeycred": false, "local_time": "2019-01-23 11:57:11.834296", "logdata": {"PASSWORD": "1111", "USERNAME": "root"}, "logtype": 6001, "node_id": "opencanary-1", "src_host": "41.139.253.2", "src_port": 36653}

How would I get agrind to also parse the logdata field?

TIA!

**Comment by rcoh** [comment]
Nested json isn't super well supported right now, but try:

'* | json | json from logdata'

You can use the fields operator to drop the other fields if you want.

**Comment by geekscrapy** [comment]
Appreciate it's not trivial, however the main usecase for me wpuld be to read JSON converted EVTX (windows logs) where there may (or may not be) multiple levels of json... Would be killer if that would be supported

**Comment by rcoh** [comment]
Yeah, makes sense. I assume you don't know the keys ahead of time? What would the ideal workflow be for you? Something like:

```* | json | count by logdata.user```?

**Comment by rcoh** [comment]
Keeping the issue open to discuss longer term improvements

**Comment by geekscrapy** [comment]
Yea, the keys relate to the event id (and there are thousands of event ids....).

Btw you probably know, but that example JSON I gave earlier is a random log I had floating around, it's not a Windows EVTX. I could provide you one if you wanted (or you can just copy one from a Windows machine).

The original format of EVTX is binary XML, which is usually extracted using the following python library. This XML is usually then converted to a python dict, then to JSON format. So it's a bit of a pain, but if you could shortcut the conversion process, that'd be a massive win. There are very few tools that allow analysis of EVTX on Linux. But maybe there is a good reason for this and I'm missing it 😂

https://github.com/williballenthin/python-evtx/issues/47

**Comment by rcoh** [comment]
yeah I'm probably not going to at specific EVTX support...I could add
support to redirect the pipeline to another program though

**Comment by geekscrapy** [comment]
No that's fine, I was more hoping for either nested XML or JSON. We only really need one of those and it'll be covered.

Adding an option to redirect seems a little more complicated than it needs to be I think. ```cat``` works fine 😀

**Comment by rcoh** [comment]
You might also try JQ to do whatever complex JSON munging you want to do
before piping the result to angle-grinder

**Comment by rcoh** [comment]
I think a good solution for this could be splatting JSON objects into 1-row-per KV like https://github.com/tomnomnom/gron -- @geekscrapy I'm curious if that would create a usable output for EVTX JSON

**Comment by geekscrapy** [comment]
Hey! This may work (disclaimer: I've not played with gron, but it sounds like it should work)

**Comment by rcoh** [comment]
With #73 adding a `splat` operator should be all that's required to get decent support for arbitrary nested structures.

**Comment by Arch-vile** [comment]
I tried running the example you gave earlier (dropped some fields here for clarity):
`echo '{"dst_host": "159.65.224.130", "logdata": {"PASSWORD": "1111", "USERNAME": "root"} }' | agrind '* | json | json from logdata | fields PASSWORD, USERNAME'`

But end up with:
`error: Expected string, found other`

```
agrind --version
ag 0.18.0
```

**Comment by rcoh** [comment]
ah, that was probably written before proper nested field support was added.

You can now refer to `logdata.PASSWORD` and `logdata.USERNAME` directly. If you want to restructure things, you can do it like this:
`* | json | logdata.PASSWORD as password | logdata.USERNAME as username | fields username, password`:
```
echo '{"dst_host": "159.65.224.130", "logdata": {"PASSWORD": "1111", "USERNAME": "root"} }' | agrind '* | json | logdata.PASSWORD as password | logdata.USERNAME as username | fields username, password'
[password=1111]            [username=root]
```
